### PR TITLE
Appbar iconTheme override fix

### LINKED ISSUE #118677: App bar leading theme overrides should apply to all widgets with IconButton children
Relates to https://github.com/flutter/flutter/pull/118216. 

This fix allowed users to supply values to `appBarTheme.iconTheme` and `appBarTheme.actionsIconTheme`, which would override any `theme.iconButtonTheme` also set in the theme. 

However, this fix only applies to widgets that are directly `IconButtons`. If, for example, our app bar was defined with these theme overrides:
```
return Theme(
      data: theme.copyWith(
        iconButtonTheme: IconButtonThemeData(
          style: IconButton.styleFrom(
            foregroundColor: Colors.yellow,
          ),
        ),
        appBarTheme: theme.appBarTheme.copyWith(
          actionsIconTheme: actionTheme!.copyWith(color: Colors.blue),
          iconTheme: leadingTheme!.copyWith(color: Colors.blue),
        ),
      ),
      child: Scaffold(
        appBar: AppBar(
          actions: [
            IconButton(
              icon: Icon(Icons.favorite),
              onPressed: () {},
            ),
            BackButton(), // this takes actionsIconTheme color value as expected
          ],
          leading: BackButton(), // this takes iconButtonTheme color value, not iconTheme
          title: Text('title'),
        ),
        body: body,
```

We get this result:
4cpMFh5rmnXsgsN

This is due to `app_bar.dart` checking for an exact widget type of `IconButton` before it applies the local theme override. So, it doesn't get applied to widgets like `BackButton`, or any other custom widget which contains an icon button.

### PR DESCRIPTION
**Pull request by esouthren**
Fixes https://github.com/flutter/flutter/issues/118677.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -1023,39 +1023,37 @@ class _AppBarState extends State<AppBar> {
     }
     if (leading != null) {
       if (theme.useMaterial3) {
-        if (leading is IconButton) {
```
Comment (QuncCccccc) [inline packages/flutter/lib/src/material/app_bar.dart:1026]: Removing this check might cause a lot of scuba failures like this: https://scuba.corp.google.com/#/report/21d4ef3c-5e05-447e-96ab-6047815fc80c?groupBy=Golden+Location&pageSize=25&pageIndex=0&selected_node=21d4ef3c-5e05-447e-96ab-6047815fc80c&status=2

Checking if the leading is a `IconButton` is originally used to fix the size of leading button size on AppBar([pr link](https://github.com/flutter/flutter/pull/110722)), so we might need to keep this but in order to make the IconTheme work, we can do

```dart
leading = IconButtonTheme(
    data: effectiveIconButtonTheme,
    child: leading is IconButton ? Center(child: leading) : leading,
);
```

**Review by QuncCccccc — COMMENTED** [review]
Thanks a lot for the fix! Just left one comment below:)

File: packages/flutter/lib/src/material/app_bar.dart (hunk 1 above)
Comment (esouthren) [reply packages/flutter/lib/src/material/app_bar.dart:1026]: Ooh great, thank you for the fix! I've added this and verified by running some of the targets in those failing scubas.

**Review by QuncCccccc — APPROVED** [review]
LGTM! Thanks:)

**Pull request merged**